metricbeat/module/redis/key/key.go

```diff
@@ -0,0 +1,115 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package key
+
+import (
+	"time"
+
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/metricbeat/mb"
+	"github.com/elastic/beats/metricbeat/mb/parse"
+	"github.com/elastic/beats/metricbeat/module/redis"
+
+	rd "github.com/garyburd/redigo/redis"
+)
+
+var (
+	debugf = logp.MakeDebug("redis-key")
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet("redis", "key", New,
+		mb.WithHostParser(parse.PassThruHostParser),
+	)
+}
+
+// MetricSet for fetching Redis server information and statistics.
+type MetricSet struct {
+	mb.BaseMetricSet
+	pool     *rd.Pool
+	patterns []KeyPattern
+}
+
+// KeyPattern contains the information required to query keys
+type KeyPattern struct {
+	Keyspace uint   `config:"keyspace"`
+	Pattern  string `config:"pattern" validate:"required"`
+	Max      uint   `config:"max"`
+}
+
+// New creates new instance of MetricSet
+func New(base mb.BaseMetricSet) (mb.MetricSet, error) {
+	// Unpack additional configuration options.
+	config := struct {
+		IdleTimeout time.Duration `config:"idle_timeout"`
+		Network     string        `config:"network"`
+		MaxConn     int           `config:"maxconn" validate:"min=1"`
+		Password    string        `config:"password"`
```

Comment:
Config and pool initialization can be probably reused in other metricsets, to be fixed in a follow up.

Reply:
Issue created for follow ups https://github.com/elastic/beats/issues/9678